src/server/frontend_xwayland/xwayland_surface.cpp

```diff
@@ -183,6 +228,19 @@ auto property_handler(
 {
     return property_handler<T>(connection, window, property, mf::XCBConnection::Handler<T>{std::move(handler)});
 }
+
+template<typename T>
+auto optional_from(bool return_value, T&& value) -> std::optional<T>
```

Comment:
I see the utility, but I'm not entirely convinced by this naming. The "optional" part of the name could be referring to the internal logic or what is built, the "from" part doesn't really suggest constructing a new object.

What I'd like is `make_optional(value, if(condition))`. But making that compile would be abusing the language.

Reply:
How about:
```c++
auto make_optional_when(bool return_value, T&& value) -> std::optional<T>
...
    make_optional_when(event->value_mask & XCB_CONFIG_WINDOW_X, geom::X{event->x}),
    make_optional_when(event->value_mask & XCB_CONFIG_WINDOW_Y, geom::Y{event->y}),
    make_optional_when(event->value_mask & XCB_CONFIG_WINDOW_WIDTH, geom::Width{event->width}),
    make_optional_when(event->value_mask & XCB_CONFIG_WINDOW_HEIGHT, geom::Height{event->height}));
```